### PR TITLE
Add member replace failpoint to robustness tests

### PR DESCRIPTION
**Pull request by serathius**
Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#contribution-flow.

### CODE REVIEW

File: tests/robustness/failpoints.go [hunk 1]
```diff
@@ -570,3 +574,123 @@ func (f killAndGofailSleep) Available(config e2e.EtcdProcessClusterConfig, membe
...
+		return err
+	}
+	for {
+		_, found, err := getID(ctx, cc, member.Config().Name)
+		if err != nil {
+			continue
+		}
+		if found {
+			break
+		}
```
Comment (ahrtr) [inline tests/robustness/failpoints.go:662]: if `found == false`, then it's a bug.

File: tests/robustness/failpoints.go [hunk 2]
```diff
@@ -570,3 +574,123 @@ func (f killAndGofailSleep) Available(config e2e.EtcdProcessClusterConfig, membe
 	}
 	return memberFailpoints.Available(f.failpoint)
 }
+
+type memberReplace struct{}
+
+func (f memberReplace) Inject(ctx context.Context, t *testing.T, lg *zap.Logger, clus *e2e.EtcdProcessCluster) error {
+	memberId := rand.Int() % len(clus.Procs)
+	member := clus.Procs[memberId]
+	endpoints := []string{}
```
Comment (ahrtr) [inline tests/robustness/failpoints.go:583]:
```suggestion
	var endpoints []string
```

**Review by ahrtr — APPROVED** [review]
LGTM with two minor comments.

File: tests/robustness/failpoints.go (hunk 1 above)
Comment (serathius) [reply tests/robustness/failpoints.go:662]: getId uses member name to identify member. Adding nor starting a member doesn't set a name. Only after member joins a cluster, it makes a proposal to set its name. So this is not a bug.

File: tests/robustness/failpoints.go [hunk 3]
```diff
@@ -570,3 +574,123 @@ func (f killAndGofailSleep) Available(config e2e.EtcdProcessClusterConfig, membe
...
+	}
+	lg.Info("Starting member", zap.String("member", member.Config().Name))
+	err = member.Start(ctx)
+	if err != nil {
+		return err
+	}
+	for {
+		_, found, err := getID(ctx, cc, member.Config().Name)
+		if err != nil {
+			continue
```
Comment (serathius) [inline tests/robustness/failpoints.go:658]: Should check ctx Done here.

File: tests/robustness/failpoint/failpoint.go [hunk 4]
```diff
@@ -570,3 +574,133 @@ func (f killAndGofailSleep) Available(config e2e.EtcdProcessClusterConfig, membe
...
+	defer cc.Close()
+	memberID, found, err := getID(ctx, cc, member.Config().Name)
+	if err != nil {
+		return err
+	}
+	if !found {
+		t.Fatal("Member not found")
+	}
+	// Need to wait health interval for cluster to accept member changes
+	time.Sleep(etcdserver.HealthInterval)
```
Comment (jmhbnz) [inline tests/robustness/failpoint/failpoint.go:605]: Do we need to have some further verification here that the member is actually operating successfully?

**Review by jmhbnz — APPROVED** [review]
LGTM - Just one question below

File: tests/robustness/failpoint/failpoint.go [hunk 5]
```diff
@@ -570,3 +574,133 @@ func (f killAndGofailSleep) Available(config e2e.EtcdProcessClusterConfig, membe
...
+	if err != nil {
+		return err
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+		_, found, err := getID(ctx, cc, member.Config().Name)
```
Comment (fuweid) [inline tests/robustness/failpoint/failpoint.go:666]: Code changes looks good. Just have one question here: do we need to wait for the sync of snapshot or logs here?

File: tests/robustness/failpoint/failpoint.go (hunk 4 above)
Comment (serathius) [reply tests/robustness/failpoint/failpoint.go:605]: No, health verification is done before failpoint injection. Member is already healthy and serving requests. Membership changes just require this additional wait.

File: tests/robustness/failpoint/failpoint.go (hunk 5 above)
Comment (serathius) [reply tests/robustness/failpoint/failpoint.go:666]: If you are asking, are we sure that restarted member is serving requests, then there is no 100% percent chance that member. After a fail point injection, we check if member is healthy and we give it a second, but that's not enough to guarantee that we record any requests after.

Comment (fuweid) [reply tests/robustness/failpoint/failpoint.go:666]: Thanks for the comment. LGTM

Comment (serathius) [reply tests/robustness/failpoint/failpoint.go:666]: No problem, this issue has been on my mind for long time. It's a good candidate for followup.

**Pull request merged**